Firmware/motion_control.cpp

```diff
@@ -25,121 +26,137 @@
+       tool precision in some cases. Therefore, arc path correction is implemented.
+
+       The small angle approximation was removed because of excessive errors for small circles (perhaps unique to
+       3d printing applications, causing significant path deviation and extrusion issues).
+       Now there will be no corrections applied, but an accurate initial sin and cos will be calculated.
+       This seems to work with a very high degree of accuracy and results in much simpler code.
 
+       Finding a faster way to approximate sin, knowing that there can be substantial deviations from the true
+       arc when using the previous approximation, would be beneficial.
+    */
+
+    // If there is only one segment, no need to do a bunch of work since this is a straight line!
+    if (segments > 1)
+    {
+        // Calculate theta per segments, and linear (z) travel per segment, e travel per segment
+        // as well as the small angle approximation for sin and cos.
+        const float theta_per_segment = angular_travel_total / segments,
+            linear_per_segment = travel_z / (segments),
+            segment_extruder_travel = (target[E_AXIS] - position[E_AXIS]) / (segments),
+            sq_theta_per_segment = theta_per_segment * theta_per_segment,
+            sin_T = theta_per_segment - sq_theta_per_segment * theta_per_segment / 6,
+            cos_T = 1 - 0.5f * sq_theta_per_segment;
+        // Loop through all but one of the segments.  The last one can be done simply
+        // by moving to the target.
+        for (uint16_t i = 1; i < segments; i++) {
+            if (n_arc_correction-- == 0) {
+                // Calculate the actual position for r_axis_x and r_axis_y
+                const float cos_Ti = cos(i * theta_per_segment), sin_Ti = sin(i * theta_per_segment);
+                r_axis_x = -offset[X_AXIS] * cos_Ti + offset[Y_AXIS] * sin_Ti;
+                r_axis_y = -offset[X_AXIS] * sin_Ti - offset[Y_AXIS] * cos_Ti;
+                // reset n_arc_correction
+                n_arc_correction = cs.n_arc_correction;
+            }
+            else {
+                // Calculate X and Y using the small angle approximation
+                const float r_axisi = r_axis_x * sin_T + r_axis_y * cos_T;
+                r_axis_x = r_axis_x * cos_T - r_axis_y * sin_T;
+                r_axis_y = r_axisi;
+            }
+
+            // Update Position
+            position[X_AXIS] = center_axis_x + r_axis_x;
+            position[Y_AXIS] = center_axis_y + r_axis_y;
+            position[Z_AXIS] += linear_per_segment;
+            position[E_AXIS] += segment_extruder_travel;
+            // Clamp to the calculated position.
+            clamp_to_software_endstops(position);
+            // Insert the segment into the buffer
+            plan_buffer_line(position[X_AXIS], position[Y_AXIS], position[Z_AXIS], position[E_AXIS], feed_rate, extruder, position);
```

Comment:
Depending on what we plan to do during recovery, the last argument to plan_buffer_line here the 4 floats are up for "abuse". In G1 this would be "target", since the final coordinate is the invariant during the move.

What we do for for arcs can be different.

Reply:
It can be different, but we still need to store XYZE since all of these could be relative. I just find it annoying that I and J is always relative as well. The only purpose of gcode_target is to fix issues with relative gcode commands

Reply:
Depending on how we see this, 4 extra floats will always be required somehow for either the (absolute) start or the end of the move because of this :/

Reply:
Can you explain how it's done differently in marlin 2.0 that could be used here (in order to save on resources)?